### PR TITLE
feat(internal): lazy git initialization

### LINKED ISSUE #6616: Lazy git clone/fetch
**What would you like Renovate to be able to do?**

Delay cloning or fetching git until it's necessary, and skip it on runs when it's not.

**Describe the solution you'd like**

The repository cache in #6589 should be expanded to incorporate config about:
- init/config
- scan/extract
- existing branch updates

With lazy git it would mean skipping the git clone/fetch altogether if:
- the repository default branch commit is unchanged and it was found to be disabled previous run
- the repository default branch commit is unchanged and no dependencies were extracted the previous run
- the existing branch updates cached on the previous run are still all "up to date"

**Describe alternatives you've considered**

Today Renovate eagerly clones/fetches and always performs a full initialization, scan and apply.

### LINKED ISSUE #6734: Cache init config based on config default branch SHA
**What would you like Renovate to be able to do?**

Skip the entire repo `init` process if the previous result was cached and the repository's default branch hasn't received a new commit.

**Did you already have any implementation ideas?**

Relies on #6733 to be finished first, and will only work for platforms that can implement that.

I think everything we look up during the repo init process should be cacheable except getting vulnerability alerts from GitHub's API, so that will need to be moved to post-init.

### PR DESCRIPTION
**Pull request by rarkins**
Leans on the cache to delay initializing the git layer. With this current implementation, onboarded repos will skip git initialization altogether if they have (a) an unchanged default branch commit, and (b) no updates/PRs to do.

Closes #6616, Closes #6734

### CODE REVIEW

**Comment by rarkins** [comment]
This current WIP successfully skips a git clone for an onboarded repo with no updates/PRs pending. There's probably a few edge cases that fail. The way we do onboarding in particular is a little strange in places and should be refactored too.

Also blocked by #6669

File: lib/workers/repository/init/index.ts [hunk 1]
```diff
@@ -11,21 +11,52 @@ import { mergeRenovateConfig } from './config';
...
-  config = await initApis(config);
+  await repositoryCache.initialize(config);
+  cache = repositoryCache.getCache();
+  cache.init = cache.init || {};
+}
+
+function validCache(config: RenovateConfig): boolean {
+  return !!(
+    config.defaultBranch === cache.init.defaultBranch &&
+    cache.init.defaultBranchSha &&
```
Comment (rarkins) [inline lib/workers/repository/init/index.ts:30]: This only works with github so far, as it returns the default branch Sha from initRepo()

Comment (viceice) [reply lib/workers/repository/init/index.ts:30]: for bitbucket server we need to fetch branches endpoint to get current sha 🙃 

https://docs.atlassian.com/bitbucket-server/rest/7.0.1/bitbucket-rest.html#idp203

File: lib/workers/repository/onboarding/branch/index.ts [hunk 2]
```diff
@@ -15,7 +16,7 @@ export async function checkOnboardingBranch(
 ): Promise<RenovateConfig> {
   logger.debug('checkOnboarding()');
   logger.trace({ config });
-  const repoIsOnboarded = await isOnboarded(config);
+  const repoIsOnboarded = config.repoIsOnboarded ?? (await isOnboarded(config));
```
Comment (rarkins) [inline lib/workers/repository/onboarding/branch/index.ts:19]: I think this is now unnecessary - will check

File: lib/workers/repository/init/index.ts [hunk 3]
```diff
@@ -12,13 +12,42 @@ import { mergeRenovateConfig } from './config';
...
 }
 
 async function initializeCaches(config: RenovateConfig): Promise<void> {
   memCache.init();
   await repositoryCache.initialize(config);
+  cache = repositoryCache.getCache();
+  cache.init = cache.init || {};
+}
+
+// istanbul ignore next
```
Comment (viceice) [inline lib/workers/repository/init/index.ts:28]: Are you sure to ignore this? Do we have a way to test it? I don't want to have a broken renovate again today 🙃

File: lib/workers/repository/init/index.ts [hunk 4]
```diff
@@ -37,7 +66,13 @@ export async function initRepo(
   let config: RenovateConfig = initializeConfig(config_);
   await initializeCaches(config);
   config = await initApis(config);
-  config = await getRepoConfig(config);
+  // istanbul ignore if
```
Comment (viceice) [inline lib/workers/repository/init/index.ts:69]: We should  better add a simple test for this. run this function twice should do it i think.

File: lib/workers/repository/init/index.ts (hunk 3 above)
Comment (rarkins) [reply lib/workers/repository/init/index.ts:28]: True, I could at least export this to unit test it

File: lib/workers/repository/init/cache.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
+import { RenovateConfig } from '../../../config';
+import { logger } from '../../../logger';
+import * as memCache from '../../../util/cache/memory';
+import * as repositoryCache from '../../../util/cache/repository';
+
+export async function initializeCaches(config: RenovateConfig): Promise<void> {
+  memCache.init();
+  await repositoryCache.initialize(config);
+  const cache = repositoryCache.getCache();
+  cache.init = cache.init || {};
```
Comment (viceice) [inline lib/workers/repository/init/cache.ts:10]:
```suggestion
```
should be done in `repositoryCache.initialize` ?

**Comment by rarkins** [comment]
This now works to defer git cloning when there is no `baseBranches` defined, such that no cloning will be required if base branch sha is unchanged and:
- repo is disabled, or
- repo is enabled but has no pending updates/branches

**Review by viceice — APPROVED** [review]

**Pull request merged**